integrations/website-indexer/src/index.ts

```diff
@@ -0,0 +1,277 @@
+     */
+    throw new RuntimeError('Invalid configuration') // replace this with your own validation logic
+  },
+  actions: {
+    indexUrls: async ({ input, logger, client }) => {
+      const pageUrlsSchema = z.object({
+        urls: z.array(z.string()),
+      })
+      const { urls: pageUrls } = pageUrlsSchema.parse(JSON.parse(input.pageUrls))
+
+      logger.forBot().debug(`Indexing ${pageUrls.length} urls`)
+
+      const scraper = new Scraper(logger, bp.secrets.SCRAPER_API_KEY)
+      let scraperCreditCost = 0
+      const files = await Promise.all(
+        pageUrls.map(async (url) => {
+          const scrapingResponse = await scraper.fetchPageHtml(url)
+          scraperCreditCost += scrapingResponse.cost
+          const response = await client.upsertFile({
+            key: url,
+            size: scrapingResponse.content.byteLength,
+            index: true,
+            contentType: 'text/html',
+          })
+          await axios.put(response.file.uploadUrl, scrapingResponse.content)
+          return response.file.id
+        })
+      )
+
+      return { fileIds: JSON.stringify(files), scraperCreditCost }
+    },
+    fetchUrls: async ({ input: { rootUrl, maxUrls } }) => {
+      const urls = await fetchUrls(rootUrl)
+      return { urls: urls.slice(0, maxUrls) }
+    },
+  },
+  channels: {},
+  handler: async () => {},
+})
+
+const fetchUrls = async (url: string): Promise<string[]> => {
+  const finalUrls = new Set<string>()
+  finalUrls.add(url)
+
+  const robotsProvided = isRobotsTxt(url)
+  const sitemapProvided = isSitemapXml(url)
+  const host = Url.parse(url)?.host ?? ''
+
+  // If the user provided a robots.txt or a sitemap.xml, we don't need to discover them
+  const sitemaps = robotsProvided
+    ? await fetchRobotsSitemapUrls(url)
+    : sitemapProvided
+    ? [url]
+    : // if the user provided a url, we need to discover the sitemaps
+      _.uniq([
+        ...(await fetchRobotsSitemapUrls(`https://${host}/robots.txt`)),
+        ...(await fetchRobotsSitemapUrls(`${url}/robots.txt`)),
+        ...getWellKnownSitemaps(url),
+        ...getWellKnownSitemaps(`https://${host}`),
+      ])
+
+  await Promise.all(
+    sitemaps.map(async (sitemap) => {
+      const entries = await fetchSitemap({ url: sitemap, depth: 3, maxUrls: 5000, sameDomain: false }).catch((err) => {
+        console.error(`Error while reading sitemap ${sitemap}: ${err?.message}`)
+        return []
+      })
+
+      entries.forEach((item) => finalUrls.add(item.url))
+    })
+  )
+
+  return [...finalUrls].filter((x) => !isRobotsTxt(x) && !isSitemapXml(x))
+}
+
+const fetchRobotsSitemapUrls = async (url: string): Promise<string[]> => {
+  try {
+    const content = await fetchPageContent(url)
+    const sitemaps = parseRobots(url, content ?? '').getSitemaps()
+    return sitemaps
+  } catch {
+    return []
+  }
+}
+
+const fetchPageContent = async (url: string) => {
+  const response = await axios.get(url)
+  return response.data
+}
```

Comment:
:bulb: **suggestion (non-blocking):** fetching the robots file should also go through the scraper because the same protection that prevents us from fetching the pages also protects the sitemap file (or at least, try fetching directly and then fallback on the scraper if it fails)

Reply:
I'll keep this in mind, thanks for suggesting!